File: python/benchmark.py
import os
import sys
import json
import time
import shutil
import numpy as np
import nibabel as nib
from fury import io

# Ensure local python directory is in import path
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from utils import (release_memory, evict_from_cache, load_data)

FILENAMES = [
    "f16_ui32_w_metadata.trx", "f16_ui32_wo_metadata.trx",
    "f16_ui64_w_metadata.trx", "f16_ui64_wo_metadata.trx",
    "f32_ui64_w_metadata.trx", "f32_ui64_wo_metadata.trx",
    "f64_ui32_w_metadata.trx", "f64_ui32_wo_metadata.trx",
    "f32_ui32_w_metadata.trx", "f32_ui32_wo_metadata.trx",
    "f64_ui64_w_metadata.trx", "f64_ui64_wo_metadata.trx",
    "f32_w_metadata.trk", "f32_wo_metadata.trk", "f32.tck",
    "f32_ui32_wo_metadata.vtk", "f32_ui64_wo_metadata.vtk",
    "f64_ui32_wo_metadata.vtk", "f64_ui64_wo_metadata.vtk",
    "f32_ui64_w_metadata.vtk", "f64_ui64_w_metadata.vtk"
]

EXPECTED_STREAMLINES = None
EXPECTED_POINTS = None

def get_counts(obj):
    if hasattr(obj, 'streamlines'):
        streamlines = obj.streamlines
        return len(streamlines), len(streamlines._data)
    elif hasattr(obj, 'GetNumberOfLines'):
        return obj.GetNumberOfLines(), obj.GetNumberOfPoints()
    else:
        raise ValueError(f"Unknown object type: {type(obj)}")

def main():
    data_dir = os.environ.get("TRX_BENCHMARK_DATA_DIR")
    if not data_dir:
        sys.stderr.write("[ERROR] Environment variable TRX_BENCHMARK_DATA_DIR is not set.\n")
        sys.exit(1)
    
    if not os.path.isdir(data_dir):
        sys.stderr.write(f"[ERROR] TRX_BENCHMARK_DATA_DIR directory '{data_dir}' does not exist.\n")
        sys.exit(1)

    results_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "results")
    os.makedirs(results_dir, exist_ok=True)

    tmp_save_dir = os.path.join(results_dir, "tmp_benchmark_saving")
    os.makedirs(tmp_save_dir, exist_ok=True)

    results = {
        "language": "python",
        "data_directory": data_dir,
        "results": {
            "loading": {},
            "saving": {}
        }
    }

    for filename in FILENAMES:
        filepath = os.path.join(data_dir, filename)
        _, ext = os.path.splitext(filename)

        if not os.path.exists(filepath):
            sys.stderr.write(f"[ERROR] File not found: {filepath}\n")
            results["results"]["loading"][filename] = None
            results["results"]["saving"][filename] = None
            continue

        print(f"Benchmarking {filename}...")
        
        # 1. Benchmark Loading
        load_times = []
        obj = None
        loading_failed = False

        for i in range(11):
            evict_from_cache(filepath)
            release_memory()

            t0 = time.time()
            try:
                obj = load_data(filepath)
                
                if hasattr(obj, 'streamlines'):
                    obj.streamlines._data = obj.streamlines._data.astype(np.float32)
                duration = time.time() - t0
                
                # Verify integrity
                streamline_count, point_count = get_counts(obj)
                global EXPECTED_STREAMLINES, EXPECTED_POINTS
                if EXPECTED_STREAMLINES is None:
                    EXPECTED_STREAMLINES = streamline_count
                    EXPECTED_POINTS = point_count
                
                if streamline_count != EXPECTED_STREAMLINES or point_count != EXPECTED_POINTS:
                    raise ValueError(
                        f"Integrity check failed: expected {EXPECTED_STREAMLINES} streamlines "
                        f"and {EXPECTED_POINTS} points, got {streamline_count} and {point_count}"
                    )
            except Exception as e:
                sys.stderr.write(f"[ERROR] Loading failed for {filename} at iteration {i}: {e}\n")
                loading_failed = True
                break

            if i == 0:
                print(f"  Load Cold Run: {duration:.4f}s")
            else:
                load_times.append(duration)
                print(f"  Load Warm Run {i}: {duration:.4f}s")

        if loading_failed:
            results["results"]["loading"][filename] = None
            results["results"]["saving"][filename] = None
            if obj is not None:
                del obj
            release_memory()
            continue

        # Save load results
        results["results"]["loading"][filename] = load_times
        avg_load = np.mean(load_times)
        std_load = np.std(load_times)
        print(f"  Summary Load: {avg_load:.4f} +/- {std_load:.4f} seconds")

        # 2. Benchmark Saving
        save_times = []
        saving_failed = False

        for i in range(11):
            release_memory()
            save_path = os.path.join(tmp_save_dir, f"tmp_save_{i}{ext}")

            t0 = time.time()
            try:
                if ext in [".trk", ".tck"]:
                    nib.streamlines.save(obj, save_path)
                elif ext in [".vtk", ".vtp", ".fib"]:
                    io.save_polydata(obj, save_path, binary=True)
                elif ext == ".trx":
                    from trx.trx_file_memmap import save as save_trx
                    save_trx(obj, save_path)
                duration = time.time() - t0
            except Exception as e:
                sys.stderr.write(f"[ERROR] Saving failed for {filename} at iteration {i}: {e}\n")
                saving_failed = True
                if os.path.exists(save_path):
                    try:
                        os.remove(save_path)
                    except Exception:
                        pass
                break

            if i == 0:
                print(f"  Save Cold Run: {duration:.4f}s")
            else:
                save_times.append(duration)
                print(f"  Save Warm Run {i}: {duration:.4f}s")

            # Clean up temp file immediately
            if os.path.exists(save_path):
                try:
                    os.remove(save_path)
                except Exception as e:
                    sys.stderr.write(f"[WARN] Failed to delete temp file {save_path}: {e}\n")

        # Cleanup loaded object
        del obj
        release_memory()

        if saving_failed:
            results["results"]["saving"][filename] = None
        else:
            results["results"]["saving"][filename] = save_times
            avg_save = np.mean(save_times)
            std_save = np.std(save_times)
            print(f"  Summary Save: {avg_save:.4f} +/- {std_save:.4f} seconds")
        
        print()

    # Final cleanup of temp directory
    try:
        shutil.rmtree(tmp_save_dir)
    except Exception:
        pass

    # Save results
    output_json = os.path.join(results_dir, "python_results.json")
    with open(output_json, "w") as f:
        json.dump(results, f, indent=4)
    print(f"Results saved to {output_json}")

if __name__ == "__main__":
    main()

File: js/benchmark.mjs
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { evictFromCache, releaseMemory, loadData, saveTRK, saveTCK, saveVTK, saveTRX } from './utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const FILENAMES = [
    "f16_ui32_w_metadata.trx", "f16_ui32_wo_metadata.trx",
    "f16_ui64_w_metadata.trx", "f16_ui64_wo_metadata.trx",
    "f32_ui64_w_metadata.trx", "f32_ui64_wo_metadata.trx",
    "f64_ui32_w_metadata.trx", "f64_ui32_wo_metadata.trx",
    "f32_ui32_w_metadata.trx", "f32_ui32_wo_metadata.trx",
    "f64_ui64_w_metadata.trx", "f64_ui64_wo_metadata.trx",
    "f32_w_metadata.trk", "f32_wo_metadata.trk", "f32.tck",
    "f32_ui32_wo_metadata.vtk", "f32_ui64_wo_metadata.vtk",
    "f64_ui32_wo_metadata.vtk", "f64_ui64_wo_metadata.vtk",
    "f32_ui64_w_metadata.vtk", "f64_ui64_w_metadata.vtk"
];

const EXPECTED_POINTS = 201521017;
const MIN_EXPECTED_STREAMLINES = 5979090;
const MAX_EXPECTED_STREAMLINES = 5979100;

function calculateMean(array) {
    if (array.length === 0) return 0;
    return array.reduce((a, b) => a + b, 0) / array.length;
}

function calculateStdDev(array, mean) {
    if (array.length === 0) return 0;
    const variance = array.map(x => Math.pow(x - mean, 2)).reduce((a, b) => a + b, 0) / array.length;
    return Math.sqrt(variance);
}

async function main() {
    const dataDir = process.env.TRX_BENCHMARK_DATA_DIR;
    if (!dataDir) {
        console.error("[ERROR] Environment variable TRX_BENCHMARK_DATA_DIR is not set.");
        process.exit(1);
    }

    if (!fs.existsSync(dataDir) || !fs.statSync(dataDir).isDirectory()) {
        console.error(`[ERROR] TRX_BENCHMARK_DATA_DIR directory '${dataDir}' does not exist.`);
        process.exit(1);
    }

    const resultsDir = path.join(path.dirname(__dirname), "results");
    if (!fs.existsSync(resultsDir)) {
        fs.mkdirSync(resultsDir, { recursive: true });
    }

    const tmpSaveDir = path.join(resultsDir, "tmp_benchmark_saving");
    if (!fs.existsSync(tmpSaveDir)) {
        fs.mkdirSync(tmpSaveDir, { recursive: true });
    }

    const results = {
        language: "javascript",
        data_directory: dataDir,
        results: {
            loading: {},
            saving: {}
        }
    };

    const args = process.argv.slice(2);
    const targetFilenames = args.length > 0 ? args : FILENAMES;

    for (const filename of targetFilenames) {
        const filepath = path.join(dataDir, filename);

        if (!fs.existsSync(filepath)) {
            console.error(`[ERROR] File not found: ${filepath}`);
            results.results.loading[filename] = null;
            results.results.saving[filename] = null;
            continue;
        }

        console.log(`Benchmarking ${filename}...`);

        const loadTimes = [];
        let loadingFailed = false;

        for (let i = 0; i < 11; i++) {
            evictFromCache(filepath);
            releaseMemory();

            await new Promise(resolve => setTimeout(resolve, 1000));

            const t0 = performance.now();
            try {
                const obj = await loadData(filepath);
                const duration = (performance.now() - t0) / 1000;

                const c1 = obj.offsetPt0.length;
                const c2 = obj.offsetPt0.length - 1;
                const countOk = (c1 >= MIN_EXPECTED_STREAMLINES && c1 <= MAX_EXPECTED_STREAMLINES) ||
                                (c2 >= MIN_EXPECTED_STREAMLINES && c2 <= MAX_EXPECTED_STREAMLINES);
                const pointCount = obj.pts.length / 3;

                if (!countOk || pointCount !== EXPECTED_POINTS) {
                    throw new Error(
                        `Integrity check failed: expected streamlines in range ` +
                        `[${MIN_EXPECTED_STREAMLINES}, ${MAX_EXPECTED_STREAMLINES}] and ${EXPECTED_POINTS} points, ` +
                        `got offsetPt0.length = ${c1} (streamlines could be ${c1} or ${c2}) and ${pointCount} points`
                    );
                }

                if (i === 0) {
                    console.log(`  Load Cold Run: ${duration.toFixed(4)}s`);
                } else {
                    loadTimes.push(duration);
                    console.log(`  Load Warm Run ${i}: ${duration.toFixed(4)}s`);
                }
            } catch (e) {
                console.error(`[ERROR] Loading failed for ${filename} at iteration ${i}: ${e.message}`);
                loadingFailed = true;
                break;
            }
        }

        releaseMemory();

        if (loadingFailed) {
            results.results.loading[filename] = null;
            results.results.saving[filename] = null;
            continue;
        }

        results.results.loading[filename] = loadTimes;
        const avgLoad = calculateMean(loadTimes);
        const stdLoad = calculateStdDev(loadTimes, avgLoad);
        console.log(`  Summary Load: ${avgLoad.toFixed(4)} +/- ${stdLoad.toFixed(4)} seconds`);

        // --- Benchmarking Saving ---
        const saveTimes = [];
        let savingFailed = false;

        let obj = null;
        try {
            obj = await loadData(filepath);
        } catch (e) {
            console.error(`[ERROR] Saving benchmark loader failed: ${e.message}`);
            savingFailed = true;
        }

        const ext = path.extname(filename).toLowerCase();

        if (!savingFailed) {
            console.log(`  Benchmarking Saving (11 iterations)...`);
            for (let i = 0; i < 11; i++) {
                releaseMemory();
                await new Promise(resolve => setTimeout(resolve, 1000));

                const savePath = path.join(tmpSaveDir, `tmp_save_${i}${ext}`);
                const t0 = performance.now();
                try {
                    if (ext === '.trx') {
                        saveTRX(savePath, obj, filename);
                    } else if (ext === '.trk') {
                        saveTRK(savePath, obj, filename);
                    } else if (ext === '.tck') {
                        saveTCK(savePath, obj);
                    } else if (ext === '.vtk') {
                        saveVTK(savePath, obj);
                    }
                    const duration = (performance.now() - t0) / 1000;

                    if (i === 0) {
                        console.log(`    Save Cold Run: ${duration.toFixed(4)}s`);
                    } else {
                        saveTimes.push(duration);
                        console.log(`    Save Warm Run ${i}: ${duration.toFixed(4)}s`);
                    }
                } catch (e) {
                    console.error(`  [ERROR] Saving failed at iteration ${i}: ${e.message}`);
                    savingFailed = true;
                    if (fs.existsSync(savePath)) {
                        try { fs.unlinkSync(savePath); } catch {}
                    }
                    break;
                }

                if (fs.existsSync(savePath)) {
                    try { fs.unlinkSync(savePath); } catch {}
                }
            }
        }

        releaseMemory();

        if (savingFailed) {
            results.results.saving[filename] = null;
        } else {
            results.results.saving[filename] = saveTimes;
            const avgSave = calculateMean(saveTimes);
            const stdSave = calculateStdDev(saveTimes, avgSave);
            console.log(`  Summary Save: ${avgSave.toFixed(4)} +/- ${stdSave.toFixed(4)} seconds\n`);
        }
    }

    try {
        if (fs.existsSync(tmpSaveDir)) {
            fs.rmSync(tmpSaveDir, { recursive: true, force: true });
        }
    } catch {}

    const outputJson = path.join(resultsDir, "js_results.json");
    fs.writeFileSync(outputJson, JSON.stringify(results, null, 4));
    console.log(`Results saved to ${outputJson}`);
}

main().catch(err => {
    console.error("Unhandled error in benchmark:", err);
    process.exit(1);
});
